interactive_templates/templates/v2/analysis/report_utils.py

```diff
@@ -1,10 +1,61 @@
 import json
 import re
+from pathlib import Path
 
 import matplotlib.dates as mdates
 import matplotlib.pyplot as plt
 
 
+BASE_DIR = Path(__file__).parents[1]
+OUTPUT_DIR = BASE_DIR / "output"
+ANALYSIS_DIR = BASE_DIR / "analysis"
+CODELIST_DIR = BASE_DIR / "codelists"
+
+
+BASE_DIR = Path(__file__).parents[1]
+OUTPUT_DIR = BASE_DIR / "output"
+ANALYSIS_DIR = BASE_DIR / "analysis"
+CODELIST_DIR = BASE_DIR / "codelists"
+
+
+def calculate_rate(df, value_col, population_col, rate_per=1000, round_rate=False):
+    """Calculates the number of events per 1,000 or passed rate_per variable of the population.
+    This function operates on the given measure table in-place, adding
+    a `rate` column.
+    Args:
+        df: A measure table.
+        value_col: The name of the numerator column in the measure table.
+        population_col: The name of the denominator column in the measure table.
+        rate_per: Value to calculate rate per
+        round: Bool indicating whether to round rate to 2dp.
+    """
+    if round_rate:
+        rate = round(df[value_col] / (df[population_col] / rate_per), 2)
+
+    else:
+        rate = df[value_col] / (df[population_col] / rate_per)
+
+    rate = rate.apply(round, ndigits=4)
```

Comment:
Is this redundant? Could supply ndigits=4 in the else statement?

Reply:
It is, i'll remove it